controller/src/main/java/org/jboss/as/controller/ControlledProcessState.java

```diff
@@ -126,8 +130,13 @@ public void setRunning() {
                 break;
             }
             synchronized (service) {
-                State newState = restartRequiredFlag ? State.RESTART_REQUIRED : State.RUNNING;
+                State newState =  restartRequiredOnStarting ? State.RESTART_REQUIRED
+                        : reloadRequiredOnStarting ? State.RELOAD_REQUIRED
+                        : restartRequiredFlag ? State.RESTART_REQUIRED : State.RUNNING;
+                newStamp = restartRequiredOnStarting || reloadRequiredOnStarting ? currentStamp : newStamp;
```

Comment:
Changing the value of newStamp instead of using a new variable makes me nervous, as this executes in a loop, so the reassignment can happen more than once. It might be fine, e.g. because of the synchronization on service, but it feels fragile / easily broken by a future maintainer.